### ISSUE TITLE
Overlapping simplices in 2D triangulation

### ISSUE THREAD

**Issue by ErlerPhilipp** [qhull/qhull#145]
Hi, I'm trying to do linear interpolation in a 2D triangulation of a point cloud, similar to Scipy nD interpolation. My small test example produces 4 overlapping triangles for 4 input points. I would expect only 2 triangles. What am I missing?

The input points roughly form a square. Only 'roughly' so that joggle is not necessary. Params are "d Qt".

Here is the code (based on https://github.com/cjkimmer/qhull_simplex_containing_point and qhull#v8.1-alpha1):
```
#include "glm/glm.hpp"
#include "glm/vec2.hpp"

#include "libqhullcpp/QhullFacet.h"
#include "libqhullcpp/QhullFacetList.h"
#include "libqhullcpp/QhullPoints.h"
#include "libqhullcpp/QhullVertex.h"
#include "libqhullcpp/QhullVertexSet.h"
#include "libqhullcpp/Qhull.h"

#include <vector>
#include <iostream>
#include <iomanip>
#include <memory>

// adapted from https://github.com/cjkimmer/qhull_simplex_containing_point/

std::vector<glm::vec2> get_points_test()
{
    std::vector<glm::vec2> points;
    points.push_back(glm::vec2(0.9f, 1.0f));
    points.push_back(glm::vec2(1.0f, -1.0f));
    points.push_back(glm::vec2(-1.0f, 1.0f));
    points.push_back(glm::vec2(-1.0f, -1.0f));

    return points;
}

std::unique_ptr<orgQhull::Qhull> triangulate(const std::vector<glm::vec2>& point_cloud, int dim = 2)
{
    // adapted from https://github.com/cjkimmer/qhull_simplex_containing_point

    // TODO: find a better way without copy
    // convert point cloud to flat vector
    std::vector<realT> point_cloud_flat;
    point_cloud_flat.reserve(point_cloud.size() * dim);
    for (const glm::vec2& point : point_cloud)
    {
        point_cloud_flat.push_back(point.x);
        point_cloud_flat.push_back(point.y);
    }

    // "d" is qdelaunay command-line switch option. You may also need "Qt" to force simplices in the
    // triangulation, but the triangulation won't be unique.
    std::string options = "d";  // Delaunay
    options += " Qt";  // force simplices
    // options += " QJ";  // joggled http://www.qhull.org/html/qh-optq.htm#QJn, not compatible with Qt
    // options += " Qbb Qc Qz Q12";  from https://github.com/scipy/scipy/blob/d5530048edef50b87e2550d6bfa41a7881a2be15/scipy/spatial/_qhull.pyx#L1808
    std::unique_ptr<orgQhull::Qhull> qhull_ptr = std::make_unique<orgQhull::Qhull>("", dim, int(point_cloud.size()), point_cloud_flat.data(), options.c_str());

    return qhull_ptr;
}

int main()
{
    QHULL_LIB_CHECK

	std::vector<glm::vec2> point_cloud = get_points_test();

    // triangulate
    int dim = 2;
    std::unique_ptr<orgQhull::Qhull> qhull = triangulate(point_cloud, dim);
    orgQhull::QhullPoints vs = qhull->points();
    std::cout << " Number of points " << vs.size() << std::endl;
    orgQhull::QhullFacetList fs = qhull->facetList();
    std::cout << " Number of simplices : " << fs.size() << std::endl;

    // loop over facets and do stuff
    for (auto f : fs) {
        orgQhull::QhullVertexSet facet_vs = f.vertices();
        std::vector< orgQhull::QhullVertex > std_vs = facet_vs.toStdVector();
        std::cout << " Facet " << f.id() << " has " << std_vs.size() << " vertices" << std::endl;
		for (auto v : std_vs) {
			std::cout << "  Vertex " << v.id() << " has coordinates " << v.point().toStdVector()[0] << ", " << v.point().toStdVector()[1] << std::endl;
		}
    }
}
```

The output is:
```
 Number of points 4
 Number of simplices : 4
 Facet 1 has 3 vertices
  Vertex 3 has coordinates 0.9, 1
  Vertex 2 has coordinates 1, -1
  Vertex 1 has coordinates -1, 1
 Facet 2 has 3 vertices
  Vertex 4 has coordinates -1, -1
  Vertex 2 has coordinates 1, -1
  Vertex 1 has coordinates -1, 1
 Facet 3 has 3 vertices
  Vertex 4 has coordinates -1, -1
  Vertex 3 has coordinates 0.9, 1
  Vertex 1 has coordinates -1, 1
 Facet 4 has 3 vertices
  Vertex 4 has coordinates -1, -1
  Vertex 3 has coordinates 0.9, 1
  Vertex 2 has coordinates 1, -1
```

**Comment by ErlerPhilipp** [comment]
After hours of stepping through QHull code, I found 2 issues.
Consider this code:
```
...
orgQhull::Qhull qhull("", dim, int(point_cloud.size()), point_cloud_flat.data(), options.c_str());
orgQhull::QhullFacetList fs = qhull->facetList();
fs.selectAll();
std::cout << " Number of simplices : " << fs.count() << std::endl;
fs.selectGood();
std::cout << " Number of good simplices : " << fs.count() << std::endl;
```

1. The actually relevant facets are 'good' triangles in the C++ interface. I need to call `fs.selectAll();` or `fs.selectGood();` before iterating through the facets.
2. orgQhull::QhullFacetList is NOT a typical STD container class. `fs.size()` counts all facets, `fs.count()` counts depending on the selected facets.

**Comment by cbbarber** [comment]
Qhull's C++ interface needs documentation badly.   Definitely for the next release. I'll try to get to it beforehand.

Did you end up with a working interface?  Any suggestions for improving the C++ interface?

**Comment by ErlerPhilipp** [comment]
Hi @cbbarber!

I'm using it and I think the results are correct and fast. The code is still the same as above, I just fall back to old-C interface for actually using the triangulation, e.g. for finding the closest facet. Thanks for QHull!

There are several points that could be improved, mostly from a programmer's usability point of view. Here some things that come to my mind from my QHull journey:
1. Usability vs performance: A nice and clean C++ interface would be much easier to use but likely a bit slower. I don't know much about the use cases but I would expect that CPU is NOT a bottleneck for most. For me, it's about deep learning and I'm GPU-bound. So, I would argue to sacrifice a bit of speed for a nicer interface.
2. C++ vs C interface: The C++ interface is not very useful on its own. For many, not even advanced features of QHull, you need to fall back to C. To make C++ useful, you need to cover most of the features.
3. QHull containers vs STD containers: People are used to STD containers and iterators. Any difference to that should be clearly documented, e.g. the .size() vs .count() as described above. I don't really see the point why custom containers are used at all. The functions could return STD containers. With move semantics, this should be very fast. I don't know about lazy evaluation but there might be something in STD or Boost.
4. C++ Documentation: I think the selectAll() and similar are never mentioned in the docu. There should be more examples and/or unit tests. In general, I found the documentation very math-heavy. If you don't really care about the details and don't want to spend time on understanding all the parameters, it's very easy to get unexpected results and errors. E.g. Qt vs QJ. When my 2D test data had 4 points and accidentally 3 were or on a line, I got an unexpected, hard-to-understand error. A trivial 3 points-example didn't work at all, if I remember correctly. I spent some time on finding the most robust parameters, which is now simply "d Qt".
5. Static, re-entrant, lib, default etc.: The different variants of the lib, which duplicate parts of the code, are quite confusing. Wouldn't it be better to have all of this in one place instead of many separate pieces. Switches with defines could be enough.
6. State-Machine semantics: selectAll() sounds like a finite state machine, which makes everything unnecessarily complicated, e.g. OpenGL which is moved towards stateless. Another bad example is the old Meshlabserver. Please don't go this direction. Just offer functions for procedural programming. Lazy evaluation could be a solution. In general, I would expect this structure: triangulate once, get requested data at once in a STD vector (e.g. all real facets).
7. 2D/3D vs nD: I expect most people are using QHull in 2D or 3D. A dedicated interface for that would be much easier to understand.
8. Double vs float: Float should be fine for me but I ran into memory issues when I tried. If I got this right, users are supposed to change the user.h and/or user.c. I can't really check in this change to my repo, so it's a fundamental problem. Usually this switch is a define or template. A template would be a nice point for the C++ interface.
9. Remove outdated stuff: There is so much old, likely useless stuff floating around. E.g. Borland projects. I guess cmake is standard nowadays. I found a strange define somewhere (ANSI C? forgot unfortunately), which dates back to when C didn't know return types.

TLDR:
1. The C++ interface needs to cover more features. My application requires Delaunay triangulation and efficiently finding the facet closest to a given point.
2. The C++ interface should be mostly procedural. Some OOP for interfaces is ok but otherwise it should be independent function calls. 
3. The C++ interface should use default STD containers. Any deviation from STD should be clearly documented.
4. More examples please. I had to use e.g. the Scipy implementation as a reference.
5. The QHull library in general should be updated to modern coding standards, e.g. no user settings in lib files, get rid of legacy stuff. Only one library? No duplicate code. If you want to keep the non-re-entrant version, make a separate branch?
6. CPU is likely no bottleneck anymore. Feel free to sacrifice some efficiency for usability.

**Comment by cbbarber** [comment]
Good advice.   The current implementation works well for me, but has difficulties for most users. A rewrite may be doable.

CGAL [www.cgal.org](http://www.cgal.org) includes a high-quality computational geometry package.  It is supported by a team of developers.  CGAL uses templates for 2D, 3D, and higher-dimensional data structures.

**Comment by ErlerPhilipp** [comment]
Yes, I know CGAL. I used it a few years ago. Quite a monster. All the template levels caused long compile times, even with pre-compiled headers. I'd like to avoid it unless necessary.

I'd be glad to give more feedback on a new version. Thanks again for your work!

**Comment by cbbarber** [comment]
Phillip,

No need to repost.  I've added your suggestions to a new wiki page for updating Qhull

https://github.com/qhull/qhull/wiki/Qhull-rework-suggestions

**Comment by ErlerPhilipp** [comment]
It seems this was too long on my TODO list. Thank you. Good to see that there is still active development.
